Repair segment retains RUNNING state following segment repair failure and Cassandra backend timeout

Running the latest 1.1.0-SNAPSHOT with Cassandra backend. I have a repair which is stuck (2 segments remaining).

The reason for that is that these two segments are in the RUNNING state according to the repair_run table:
```

cqlsh> select segment_id from reaper_db.repair_run where id = 40aee550-0069-11e8-9208-931ae1c16c5b and segment_state = 1 allow filtering

 segment_id
--------------------------------------
 40aee59c-0069-11e8-9208-931ae1c16c5b
 40b0451b-0069-11e8-9208-931ae1c16c5b

```
 
However there are no segment runners for them. Hence when `RepairRunner.startNextSegment` runs it logs `Running segment for range` and then `context.storage.getNextFreeSegmentInRange(repairRunId, Optional.of(parallelRanges.get(rangeIndex)))` does not return any segment because it returns segments in the NOT_STARTED state only.

According to the logs this is what happened to one of these segments stuck in RUNNING state:

1.  Segment repair (40aee59c-0069-11e8-9208-931ae1c16c5b) gets started:
`INFO   [2018-01-25 11:30:20,840] [paidrecs:40aee550-0069-11e8-9208-931ae1c16c5b] i.c.s.RepairRunner - Next segment to run : 40aee59c-0069-11e8-9208-931ae1c16c5b
INFO   [2018-01-25 11:30:52,705] [paidrecs:40aee550-0069-11e8-9208-931ae1c16c5b:40aee59c-0069-11e8-9208-931ae1c16c5b] i.c.s.SegmentRunner - It is ok to repair segment '40aee59c-0069-11e8-9208-931ae1c16c5b' on repair run with id '40aee550-0069-11e8-9208-931ae1c16c5b'
`
2. Segment repair fails and backend update times out:
`
ERROR  [2018-01-25 11:31:24,082] [paidrecs:40aee550-0069-11e8-9208-931ae1c16c5b:40aee59c-0069-11e8-9208-931ae1c16c5b] i.c.s.RepairRunner - Executing SegmentRunner failed
com.datastax.driver.core.exceptions.OperationTimedOutException: [/10.40.66.47:9042] Timed out waiting for server response
        at com.datastax.driver.core.exceptions.OperationTimedOutException.copy(OperationTimedOutException.java:44)
        at com.datastax.driver.core.exceptions.OperationTimedOutException.copy(OperationTimedOutException.java:26)
        at com.datastax.driver.core.DriverThrowables.propagateCause(DriverThrowables.java:37)
        at com.datastax.driver.core.DefaultResultSetFuture.getUninterruptibly(DefaultResultSetFuture.java:245)
        at com.datastax.driver.core.AbstractSession.execute(AbstractSession.java:68)
        at io.cassandrareaper.storage.CassandraStorage.hasLeadOnSegment(CassandraStorage.java:1046)
        at io.cassandrareaper.storage.CassandraStorage.updateRepairSegment(CassandraStorage.java:618)
        at io.cassandrareaper.service.SegmentRunner.processTriggeredSegment(SegmentRunner.java:299)
        at io.cassandrareaper.service.SegmentRunner.runRepair(SegmentRunner.java:277)
        at io.cassandrareaper.service.SegmentRunner.run(SegmentRunner.java:133)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:108)
        at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:41)
        at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:77)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: com.datastax.driver.core.exceptions.OperationTimedOutException: [/10.40.66.47:9042] Timed out waiting for server response
        at com.datastax.driver.core.RequestHandler$SpeculativeExecution.onTimeout(RequestHandler.java:772)
        at com.datastax.driver.core.Connection$ResponseHandler$1.run(Connection.java:1374)
        at io.netty.util.HashedWheelTimer$HashedWheelTimeout.expire(HashedWheelTimer.java:581)
        at io.netty.util.HashedWheelTimer$HashedWheelBucket.expireTimeouts(HashedWheelTimer.java:656)
        at io.netty.util.HashedWheelTimer$Worker.run(HashedWheelTimer.java:367)
`

Since then there is nothing about 40aee59c-0069-11e8-9208-931ae1c16c5b segment in the logs - and the segment is left in the RUNNING state in the storage without a segment runner instance for it.

--- a/src/server/src/main/java/io/cassandrareaper/service/RepairManager.java
+++ b/src/server/src/main/java/io/cassandrareaper/service/RepairManager.java
@@ -34,6 +34,7 @@ import java.util.stream.Collectors;
 import com.codahale.metrics.MetricRegistry;
 import com.codahale.metrics.Timer;
 import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Optional;
 import com.google.common.base.Preconditions;
 import com.google.common.collect.Maps;
 import com.google.common.util.concurrent.ListenableFuture;
@@ -91,39 +92,49 @@ public final class RepairManager {
    * @param context Reaper's application context.
    */
   public void resumeRunningRepairRuns() throws ReaperException {
-    heart.beat();
-    Collection<RepairRun> running = context.storage.getRepairRunsWithState(RepairRun.RunState.RUNNING);
-    for (RepairRun repairRun : running) {
+    try {
+      heart.beat();
+      Collection<RepairRun> running =
+          context.storage.getRepairRunsWithState(RepairRun.RunState.RUNNING);
+      for (RepairRun repairRun : running) {
 
-      Collection<RepairSegment> runningSegments
-          = context.storage.getSegmentsWithState(repairRun.getId(), RepairSegment.State.RUNNING);
+        Collection<RepairSegment> runningSegments =
+            context.storage.getSegmentsWithState(repairRun.getId(), RepairSegment.State.RUNNING);
 
-      abortSegmentsWithNoLeader(repairRun, runningSegments);
+        abortSegmentsWithNoLeader(repairRun, runningSegments);
 
-      if (!repairRunners.containsKey(repairRun.getId())) {
-        LOG.info("Restarting run id {} that has no runner", repairRun.getId());
-        startRepairRun(repairRun);
-      }
-    }
-
-    Collection<RepairRun> paused = context.storage.getRepairRunsWithState(RepairRun.RunState.PAUSED);
-    for (RepairRun pausedRepairRun : paused) {
-      if (repairRunners.containsKey(pausedRepairRun.getId())) {
-        // Abort all running segments for paused repair runs
-        Collection<RepairSegment> runningSegments
-            = context.storage.getSegmentsWithState(pausedRepairRun.getId(), RepairSegment.State.RUNNING);
-
-        abortSegments(runningSegments, pausedRepairRun);
+        if (!repairRunners.containsKey(repairRun.getId())) {
+          LOG.info("Restarting run id {} that has no runner", repairRun.getId());
+          startRepairRun(repairRun);
+        }
       }
 
-      if (!repairRunners.containsKey(pausedRepairRun.getId())) {
-        startRunner(pausedRepairRun.getId());
+      Collection<RepairRun> paused =
+          context.storage.getRepairRunsWithState(RepairRun.RunState.PAUSED);
+      for (RepairRun pausedRepairRun : paused) {
+        if (repairRunners.containsKey(pausedRepairRun.getId())) {
+          // Abort all running segments for paused repair runs
+          Collection<RepairSegment> runningSegments =
+              context.storage.getSegmentsWithState(
+                  pausedRepairRun.getId(), RepairSegment.State.RUNNING);
+
+          abortSegments(runningSegments, pausedRepairRun, false);
+        }
+
+        if (!repairRunners.containsKey(pausedRepairRun.getId())) {
+          startRunner(pausedRepairRun.getId());
+        }
       }
+    } catch (RuntimeException e) {
+      throw new ReaperException(e);
     }
   }
 
   private void abortSegmentsWithNoLeader(RepairRun repairRun, Collection<RepairSegment> runningSegments) {
 
+    LOG.debug(
+        "Checking leadership on the following segments : {}",
+        runningSegments.stream().map(seg -> seg.getId()).collect(Collectors.toList()));
     if (context.storage instanceof IDistributedStorage || !repairRunners.containsKey(repairRun.getId())) {
       // When multiple Reapers are in use, we can get stuck segments when one instance is rebooted
       // Any segment in RUNNING state but with no leader should be killed
@@ -132,19 +143,43 @@ public final class RepairManager {
               ? ((IDistributedStorage) context.storage).getLeaders()
               : Collections.emptyList();
 
+      LOG.debug(
+          "No leader on the following segments : {}",
+          runningSegments
+              .stream()
+              .filter(segment -> !activeLeaders.contains(segment.getId()))
+              .map(seg -> seg.getId())
+              .collect(Collectors.toSet()));
+
       abortSegments(
           runningSegments
               .stream()
               .filter(segment -> !activeLeaders.contains(segment.getId()))
               .collect(Collectors.toSet()),
-          repairRun);
+          repairRun,
+          true);
     }
   }
 
-  void abortSegments(Collection<RepairSegment> runningSegments, RepairRun repairRun) {
+  void abortSegments(
+      Collection<RepairSegment> runningSegments,
+      RepairRun repairRun) {
+
+    abortSegments(
+        runningSegments,
+        repairRun,
+        false);
+  }
+
+  void abortSegments(
+      Collection<RepairSegment> runningSegments,
+      RepairRun repairRun,
+      boolean postponeWithoutAborting) {
 
     RepairUnit repairUnit = context.storage.getRepairUnit(repairRun.getRepairUnitId()).get();
     for (RepairSegment segment : runningSegments) {
+      LOG.debug(
+          "Trying to abort stuck segment {} in repair run {}", segment.getId(), repairRun.getId());
       UUID leaderElectionId = repairUnit.getIncrementalRepair() ? repairRun.getId() : segment.getId();
       if (takeLead(context, leaderElectionId) || renewLead(context, leaderElectionId)) {
         // refresh segment once we're inside leader-election
@@ -152,14 +187,19 @@ public final class RepairManager {
         if (RepairSegment.State.RUNNING == segment.getState()) {
           try (JmxProxy jmxProxy = context.jmxConnectionFactory.connect(
               segment.getCoordinatorHost(), context.config.getJmxConnectionTimeoutInSeconds())) {
-
-            SegmentRunner.abort(context, segment, jmxProxy);
+            LOG.warn(
+                "Aborting stuck segment {} in repair run {}", segment.getId(), repairRun.getId());
+            if (postponeWithoutAborting) {
+              SegmentRunner.abort(context, segment, jmxProxy);
+            }
           } catch (ReaperException | NumberFormatException | InterruptedException e) {
             LOG.debug(
                 "Tried to abort repair on segment {} marked as RUNNING, "
-                    + "but the host was down  (so abortion won't be needed)",
+                    + "but the host was down  (so abortion won't be needed). "
+                    + "Postponing the segment.",
                 segment.getId(),
                 e);
+            SegmentRunner.postpone(context, segment, Optional.fromNullable(repairUnit));
           } finally {
             // if someone else does hold the lease, ie renewLead(..) was true,
             // then their writes to repair_run table and any call to releaseLead(..) will throw an exception

--- a/src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java
+++ b/src/server/src/main/java/io/cassandrareaper/service/SegmentRunner.java
@@ -144,7 +144,7 @@ final class SegmentRunner implements RepairStatusHandler, Runnable {
     }
   }
 
-  private static void postpone(AppContext context, RepairSegment segment, Optional<RepairUnit> repairUnit) {
+  static void postpone(AppContext context, RepairSegment segment, Optional<RepairUnit> repairUnit) {
     LOG.info("Postponing segment {}", segment.getId());
     try {
       context.storage.updateRepairSegment(

--- a/src/server/src/test/java/io/cassandrareaper/service/RepairManagerTest.java
+++ b/src/server/src/test/java/io/cassandrareaper/service/RepairManagerTest.java
@@ -98,7 +98,9 @@ public final class RepairManagerTest {
 
     context.repairManager.repairRunners.put(run.getId(), mock(RepairRunner.class));
 
-    Mockito.doNothing().when(context.repairManager).abortSegments(any(), any());
+    Mockito.doNothing()
+        .when(context.repairManager)
+        .abortSegments(any(), any(), Mockito.anyBoolean());
     Mockito.doReturn(run).when(context.repairManager).startRepairRun(run);
     when(context.storage.getRepairRunsWithState(RepairRun.RunState.RUNNING)).thenReturn(Arrays.asList(run));
     when(context.storage.getRepairRunsWithState(RepairRun.RunState.PAUSED)).thenReturn(Collections.emptyList());
@@ -108,7 +110,8 @@ public final class RepairManagerTest {
     context.repairManager.resumeRunningRepairRuns();
 
     // Check that abortSegments was invoked is at least one segment, meaning abortion occurs
-    Mockito.verify(context.repairManager, Mockito.times(1)).abortSegments(Mockito.argThat(new NotEmptyList()), any());
+    Mockito.verify(context.repairManager, Mockito.times(1))
+        .abortSegments(Mockito.argThat(new NotEmptyList()), any(), Mockito.anyBoolean());
   }
 
   /**
@@ -164,6 +167,9 @@ public final class RepairManagerTest {
     context.repairManager.repairRunners.put(run.getId(), mock(RepairRunner.class));
 
     Mockito.doNothing().when(context.repairManager).abortSegments(any(), any());
+    Mockito.doNothing()
+        .when(context.repairManager)
+        .abortSegments(any(), any(), Mockito.anyBoolean());
     Mockito.doReturn(run).when(context.repairManager).startRepairRun(run);
     when(context.storage.getRepairRunsWithState(RepairRun.RunState.RUNNING)).thenReturn(Arrays.asList(run));
     when(context.storage.getRepairRunsWithState(RepairRun.RunState.PAUSED)).thenReturn(Collections.emptyList());
@@ -173,7 +179,8 @@ public final class RepairManagerTest {
     context.repairManager.resumeRunningRepairRuns();
 
     // Check that abortSegments was invoked with an empty list, meaning no abortion occurs
-    Mockito.verify(context.repairManager, Mockito.times(1)).abortSegments(Mockito.argThat(new EmptyList()), any());
+    Mockito.verify(context.repairManager, Mockito.times(1))
+        .abortSegments(Mockito.argThat(new EmptyList()), any(), Mockito.anyBoolean());
   }
 
   /**
@@ -226,7 +233,9 @@ public final class RepairManagerTest {
 
     context.repairManager.repairRunners.put(run.getId(), mock(RepairRunner.class));
 
-    Mockito.doNothing().when(context.repairManager).abortSegments(any(), any());
+    Mockito.doNothing()
+        .when(context.repairManager)
+        .abortSegments(any(), any(), Mockito.anyBoolean());
     Mockito.doReturn(run).when(context.repairManager).startRepairRun(run);
     when(context.storage.getRepairRunsWithState(RepairRun.RunState.RUNNING)).thenReturn(Arrays.asList(run));
     when(context.storage.getRepairRunsWithState(RepairRun.RunState.PAUSED)).thenReturn(Collections.emptyList());
@@ -235,7 +244,8 @@ public final class RepairManagerTest {
     context.repairManager.resumeRunningRepairRuns();
 
     // Check that abortSegments was not invoked at all, meaning no abortion occurs
-    Mockito.verify(context.repairManager, Mockito.times(0)).abortSegments(any(), any());
+    Mockito.verify(context.repairManager, Mockito.times(0))
+        .abortSegments(any(), any(), Mockito.anyBoolean());
   }
 
   /**
@@ -286,7 +296,9 @@ public final class RepairManagerTest {
             .withRunId(run.getId())
             .build(UUIDs.timeBased());
 
-    Mockito.doNothing().when(context.repairManager).abortSegments(any(), any());
+    Mockito.doNothing()
+        .when(context.repairManager)
+        .abortSegments(any(), any(), Mockito.anyBoolean());
     Mockito.doReturn(run).when(context.repairManager).startRepairRun(run);
     when(context.storage.getRepairRunsWithState(RepairRun.RunState.RUNNING)).thenReturn(Arrays.asList(run));
     when(context.storage.getRepairRunsWithState(RepairRun.RunState.PAUSED)).thenReturn(Collections.emptyList());
@@ -295,7 +307,8 @@ public final class RepairManagerTest {
     context.repairManager.resumeRunningRepairRuns();
 
     // Check that abortSegments was invoked with an non empty list, meaning abortion occurs
-    Mockito.verify(context.repairManager, Mockito.times(1)).abortSegments(Mockito.argThat(new NotEmptyList()), any());
+    Mockito.verify(context.repairManager, Mockito.times(1))
+        .abortSegments(Mockito.argThat(new NotEmptyList()), any(), Mockito.anyBoolean());
   }
 
   @Test

